### PR TITLE
Comm Factory

### PR DESCRIPTION
**Pull request by phlptp**
Add a commFactory to use in the MultiBroker, and eventually in some new core types, and add a few tests of the new factory

### Summary

If merged this pull request will add a CommFactory for use with the MultiBroker and possibly other components.  

### Proposed changes

- CommFactory addition
- load the commFactory with the loadCore code 
- test the new factory in the network tests

### CODE REVIEW

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/network/CommsInterface.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5388827&bid=17667357)

File: src/helics/network/CommsInterface.hpp [hunk 1]
```diff
@@ -173,6 +173,52 @@ class CommsInterface {
...
+            "Type does not inherit from helics::CommsInterface");
+
+        using comm_build_type = CommTYPE;
+        virtual std::unique_ptr<CommsInterface> build() override
+        {
+            return std::make_unique<CommTYPE>();
+        }
+    };
+
+    //** define a new Core Builder from the builder give a name and build code*/
```
Comment (nightlark) [inline src/helics/network/CommsInterface.hpp:199]:
```suggestion
    /** define a new Core Builder from the builder give a name and build code*/
```

**Comment by phlptp** [comment]
Possibly,  Not 100% sure what the "public" API for new core types would look like.  It is part of the Core API, which for the most part isn't what we consider the Public API.  

If we go along the lines of a plugin notion for new cores, then the interface definition would need to be defined in C, and that would have to be public in some ways.    I know it certainly isn't supposed to be part of the application API interface right now.    

But it is a step towards user defined Cores.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/network/TcpSSCore-tests.cpp  1
- src/helics/network/CommsInterface.cpp  1
- tests/helics/network/ZeromqSSCore-tests.cpp  3
- tests/helics/network/ZeromqCore-tests.cpp  2
- tests/helics/network/IPCcore_tests.cpp  2
         

Clones removed
==============
+ tests/helics/network/TcpCore-tests.cpp  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5388827&bid=17667357)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/network/IPCcore_tests.cpp  2
- tests/helics/network/ZeromqSSCore-tests.cpp  3
- tests/helics/network/ZeromqCore-tests.cpp  2
- tests/helics/network/TcpSSCore-tests.cpp  3
- src/helics/network/CommsInterface.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5388827&bid=17667357)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/network/TcpSSCore-tests.cpp  3
- tests/helics/network/IPCcore_tests.cpp  2
- tests/helics/network/ZeromqCore-tests.cpp  2
- tests/helics/network/ZeromqSSCore-tests.cpp  3
- src/helics/network/CommsInterface.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5388827&bid=17667357)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/network/IPCcore_tests.cpp  2
- src/helics/network/CommsInterface.cpp  1
- tests/helics/network/ZeromqSSCore-tests.cpp  3
- tests/helics/network/ZeromqCore-tests.cpp  2
- tests/helics/network/TcpSSCore-tests.cpp  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5388827&bid=17667357)

**Pull request merged**